src/main/scala/viper/gobra/util/ViperChopper.scala

```diff
@@ -182,21 +181,26 @@ object ViperChopper {
     }
 
     sealed trait MergePenalty[T] {
-      type Rep
-      type Rev
-      def init(x: Vector[Set[T]]): (Vector[Rep], Rev)
-      def finish(y: Vector[Rep], m: Rev): Vector[Set[T]]
-      def penaltyAndMerge(l: Rep, r: Rep): (Int, Rep)
+      type Node
+      type Program
+      type Mapping
+
+      def init(nodes: Vector[T], forest: Vector[Edge[T]]): (Vector[Node], Vector[Edge[Node]], Mapping)
+      def inverse(node: Node, m: Mapping): T
+      def toProgram(set: Set[Node]): Program
+      def finish(y: Vector[Program], m: Mapping): Vector[Set[T]]
+      def penaltyAndMerge(l: Program, r: Program): (Int, Program)
     }
 
     object Penalty {
 
-      object Default extends MergePenalty[SCC.Component[Vertex]] {
+      trait DefaultImpl extends MergePenalty[SCC.Component[Vertex]] {
```

Comment:
Nitpick: even though this trait contains the entire default implementation of `MergePenalty`, it might be a bit misleading to have its name ending in `Impl`. I believe this suffix is most commonly used in names of classes and objects, not so much in traits.

Reply:
I am open to name suggestions.

Reply:
I would name the trait `Default` and the object that extends it `DefaultImpl`

Reply:
I now made the trait a class